### PR TITLE
History Window

### PR DESCRIPTION
**Pull request by Mink80**
with friendly greetings from ronin of UO SAGAS

### CODE REVIEW

**Review by 3HMonkey — APPROVED** [review]

**Review by 3HMonkey — APPROVED** [review]

**Review by 3HMonkey — APPROVED** [review]

**Comment by 3HMonkey** [comment]
Still a bug inside. Do not merge for the moment. Ill check it.

**Review by 3HMonkey — CHANGES_REQUESTED** [review]
You'll need to wait for the table to setup first when rendering its content, so you have to use its' content inside the clause. 

Example:
```csharp
if (ImGui.BeginTable("UndoTable", 2, ImGuiTableFlags.Borders | ImGuiTableFlags.SizingFixedFit))
{
    // Your doing here
    ImGui.EndTable();
}
```

So please replace:

```csharp
// Create table
ImGui.BeginTable("UndoTable", 2, ImGuiTableFlags.Borders | ImGuiTableFlags.SizingFixedFit);

// Set column widths
ImGui.TableSetupColumn("Task", ImGuiTableColumnFlags.WidthFixed, 120);
ImGui.TableSetupColumn("Details", ImGuiTableColumnFlags.WidthFixed, 350);

// Headers
ImGui.TableHeadersRow();

var cnt = 0;
foreach (var command in CEDClient.UndoStack)
{
    cnt++;
    ImGui.TableNextRow();
    ImGui.TableNextColumn();
    GetHistory(command[0], out var task, out var details);
    ImGui.Text(task);
    ImGui.TableNextColumn();
    ImGui.Text(details);

    if (cnt >= 25)
    {
        break;
    }
}
ImGui.EndTable();
```

with

```csharp
if (ImGui.BeginTable("UndoTable", 2, ImGuiTableFlags.Borders | ImGuiTableFlags.SizingFixedFit))
{
    // Set column widths
    ImGui.TableSetupColumn("Task", ImGuiTableColumnFlags.WidthFixed, 120);
    ImGui.TableSetupColumn("Details", ImGuiTableColumnFlags.WidthFixed, 350);

    // Headers
    ImGui.TableHeadersRow();

    var cnt = 0;
    foreach (var command in CEDClient.UndoStack)
    {
        cnt++;
        ImGui.TableNextRow();
        ImGui.TableNextColumn();
        GetHistory(command[0], out var task, out var details);
        ImGui.Text(task);
        ImGui.TableNextColumn();
        ImGui.Text(details);

        if (cnt >= 25)
        {
            break;
        }
    }
    ImGui.EndTable();
}
```

**Comment by Mink80** [comment]
Thank you @3HMonkey. Done.

**Review by 3HMonkey — APPROVED** [review]
Approved and runs. Good thing!

File: CentrED/UI/Windows/HistoryWindow.cs [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+                        ImGui.TableNextRow();
+                        ImGui.TableNextColumn();
+                        GetHistory(packet, out var task, out var details);
+                        ImGui.Text($"\t {task}");
+                        ImGui.TableNextColumn();
+                        ImGui.Text(details);
+                    }
+                }
+
+                if (cnt >= 25)
```
Comment (kaczy93) [inline CentrED/UI/Windows/HistoryWindow.cs:65]: Why do we limit to only 25 entries? ImGUI is very good at providing a scrolling interface if stuff wouldn't fit inside the window and if this is due to performance we can use the imgui list clipper

File: CentrED/UI/Windows/HistoryWindow.cs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            ImGui.EndPopup();
+        }
+        
+        ImGui.End();
+        
+    }
+
+    private static void GetHistory(Packet p, out string command, out string details)
+    {
+        details = "Error parsing packet"; // should never be shown
```
Comment (kaczy93) [inline CentrED/UI/Windows/HistoryWindow.cs:107]: This should be in the `default` block :)

File: CentrED/UI/Windows/HistoryWindow.cs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            // Headers
+            ImGui.TableHeadersRow();
+
+            var cnt = 0;
+            foreach (var command in CEDClient.UndoStack)
+            {
+                cnt++;
+                ImGui.TableNextRow();
+                
+                if (command.Length == 1)
```
Comment (kaczy93) [inline CentrED/UI/Windows/HistoryWindow.cs:40]: I think we can make it nicer if we don't repeat the code.
Something like this
```csharp
var isGroup = command.Length > 1
if (isGroup){
    ImGui.TableNextRow();
    ImGui.TableNextColumn();
    ImGui.Text("Task collection");
}
foreach (var packet in command)
{
    ImGui.TableNextRow();
    ImGui.TableNextColumn();
    GetHistory(packet, out var task, out var details);
    ImGui.Text($"{isGroup ? "\t " : ""}{task}");
    ImGui.TableNextColumn();
    ImGui.Text(details);
}
```

File: Client/Packets.cs [hunk 4]
```diff
@@ -88,12 +97,21 @@ public DrawMapPacket(ushort x, ushort y, sbyte z, ushort tileId) : base(0x06, 8)
...
     public InsertStaticPacket(StaticTile tile) : this(tile.X, tile.Y, tile.Z, tile.Id, tile.Hue)
     {
     }
 
     public InsertStaticPacket(ushort x, ushort y, sbyte z, ushort tileId, ushort hue) : base(0x07, 10)
     {
+        X = x;
+        Y = y;
+        Z = z;
+        TileId = tileId;
```
Comment (kaczy93) [inline Client/Packets.cs:114]: It feels weird that we omit hue :)
Also applies to other static packets

**Pull request merged**